<a href="https://colab.research.google.com/github/DarekGit/FACES_DNN/blob/master/notebooks/06_00_Porownanie.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---

[Spis treści](https://github.com/DarekGit/FACES_DNN/blob/master/notebooks/Praca_Dyplomowa.ipynb) | [1. Wstęp](01_00_Wstep.ipynb) | [2. Metryki oceny detekcji](02_00_Miary.ipynb) | [3. Bazy danych](03_00_Datasety.ipynb) | [4. Przegląd metod detekcji](04_00_Modele.ipynb) | [5. Detekcja twarzy z wykorzystaniem wybranych architektur GSN](05_00_Modyfikacje.ipynb) | [6. Porównanie modeli](06_00_Porownanie.ipynb) | [7. Eksport modelu](07_00_Eksport_modelu.ipynb) | [8. Podsumowanie i wnioski](08_00_Podsumowanie.ipynb) | [Bibliografia](Bibliografia.ipynb)


---

# 6. Porównanie modeli

Pomiary czasu interferencji wybranych modeli, wyliczenie metryk mAP oraz Coco AP wykonano dla zestawów danych FDDB, WIDER FACE oraz FACES_DD. 
Testy zostały zamieszczone w plikach: 

 * 6.1. [Wyniki na zbiorze FDDB](06_01_FDDB_TEST.ipynb)  
 * 6.2. [Wyniki na zbiorze WIDER FACE](06_02_WIDERFACE_TEST.ipynb) 
 * 6.3. [Wyniki na zbiorze FACES_DD](06_03_FACES_DD_TEST.ipynb)   

<br>

### Porównanie wyników testów


Załadowania wyników testów modeli.

In [ ]:
from google.colab import drive
import pandas as pd
import os
import json

drive.mount('/content/drive')

time_measurement= "time_measurement_results.json"
unified_mAP = "unified_mAP_dataframe.json"
COCO_AP_FDDB = "df_AP_FDDB.json"
COCO_AP_WIDERFACE = "df_AP_WIDERFACE.json"
COCO_AP_FACES_DD = "df_AP_FACES_DD.json"

unified_mAP_FDDB = os.path.join("./drive/My Drive", "FDDB_OUTPUT_DIR", unified_mAP)
unified_mAP_WIDERFACE = os.path.join("./drive/My Drive", "WIDERFACE_OUTPUT_DIR", unified_mAP)
unified_mAP_FACES_DD = os.path.join("./drive/My Drive", "FACES_DD_OUTPUT_DIR", unified_mAP)

time_measurement_FDDB = os.path.join("./drive/My Drive", "FDDB_OUTPUT_DIR", time_measurement)
time_measurement_WIDERFACE = os.path.join("./drive/My Drive", "WIDERFACE_OUTPUT_DIR", time_measurement)
time_measurement_FACES_DD = os.path.join("./drive/My Drive", "FACES_DD_OUTPUT_DIR", time_measurement)

AP_FDDB = os.path.join("./drive/My Drive", "FDDB_OUTPUT_DIR", COCO_AP_FDDB)
AP_WIDERFACE = os.path.join("./drive/My Drive", "WIDERFACE_OUTPUT_DIR", COCO_AP_WIDERFACE )
AP_FACES_DD = os.path.join("./drive/My Drive", "FACES_DD_OUTPUT_DIR", COCO_AP_FACES_DD)


In [3]:
#time
with open(time_measurement_FDDB) as data_file:
    data_loaded = json.load(data_file)
df_time_FDDB = pd.read_json(data_loaded, orient="table" ).sort_values(by=['total_time'], ascending=True)

with open(time_measurement_WIDERFACE) as data_file:
    data_loaded = json.load(data_file)
df_time_WIDERFACE = pd.read_json(data_loaded, orient="table" ).sort_values(by=['total_time'], ascending=True)

with open(time_measurement_FACES_DD) as data_file:
    data_loaded = json.load(data_file)
df_time_FACES_DD= pd.read_json(data_loaded, orient="table" ).sort_values(by=['total_time'], ascending=True)

In [86]:
# unified_mAP
with open(unified_mAP_FDDB) as data_file:
    data_loaded = json.load(data_file)
df_mAP_FDDB = pd.read_json(data_loaded, orient="table" )[['mAP', 'All 0.50']].sort_values(by=['All 0.50'], ascending=False)*100

with open(unified_mAP_WIDERFACE) as data_file:
    data_loaded = json.load(data_file)
df_mAP_WIDERFACE = pd.read_json(data_loaded, orient="table" )[['mAP', 'All 0.50']].sort_values(by=['All 0.50'], ascending=False)*100

with open(unified_mAP_FACES_DD) as data_file:
    data_loaded = json.load(data_file)
df_mAP_FACES_DD= pd.read_json(data_loaded, orient="table" ).sort_values(by=['mAP'], ascending=False)

In [88]:
# COCO AP
with open(AP_FDDB) as data_file:
    data_loaded = json.load(data_file)
df_AP_FDDB = pd.read_json(data_loaded, orient="table" )[['AP', 'AP50']].sort_values(by=['AP50'], ascending=False)

with open(AP_WIDERFACE) as data_file:
    data_loaded = json.load(data_file)
df_AP_WIDERFACE = pd.read_json(data_loaded, orient="table" )[['AP', 'AP50']].sort_values(by=['AP50'], ascending=False)


with open(AP_FACES_DD) as data_file:
    data_loaded = json.load(data_file)
df_AP_FACES_DD= pd.read_json(data_loaded, orient="table" )[['AP', 'AP50']].sort_values(by=['AP50'], ascending=False)

### Wyniki pomiaru czasu interferencji 

Pomiary czasu inferencji wykonano na Colabie z wykorzystaniem karty graficznej NVidia Tesla P100 16GB. Na datasecie FDDB najlepsze wyniki z wytrenowanych modeli osiągnął model '800k' z wynikiem FPS na poziomie 15.9. Najsłabiej wypadły modele z ResNet-50 (12.1 i 11.2 PFS). W przypadku datasetów zawierających zdjęcia o wiekszej rozdzielczości (WIDER FACE, FACES_DD) najlepsze wyniki osiągnał model 'BN_800k'. Można zauważyć również poprawę wyników modeli z ResNet-50 FPN. NAjsłąbiej w tym porównaniu wypadł detektor MTCNN.

Otrzymane wyniki zostały zaprezentowane w poniższych tabelach.

Wynik dla FDDB

In [6]:
df_time_FDDB

total_time  mean_diff        fps
MTCNN                             91.830989   0.032278  30.980827
800k                             178.880448   0.062875  15.904477
BN_800k                          179.575302   0.063120  15.842936
BN_Mish_V3_80+_30k               200.774246   0.070571  14.170144
BN_Mish_V2_250+F_2_50k           200.794809   0.070578  14.168693
faster_rcnn_R_50_FPN_3x          234.852265   0.082549  12.113999
scratch_faster_rcnn_R_50_FPN_gn  253.903901   0.089246  11.205027

Wynik dla WIDER FACE

In [7]:
df_time_WIDERFACE

total_time  mean_diff        fps
BN_800k                          244.897014   0.075914  13.172884
800k                             248.565521   0.077051  12.978469
faster_rcnn_R_50_FPN_3x          307.453097   0.095305  10.492657
BN_Mish_V3_80+_30k               311.967525   0.096704  10.340820
BN_Mish_V2_250+F_2_50k           314.006855   0.097336  10.273661
MTCNN                            329.380084   0.102102   9.794156
scratch_faster_rcnn_R_50_FPN_gn  332.273322   0.102999   9.708875

Wynik dla FACES_DD

In [56]:
df_time_WIDERFACE

total_time  mean_diff        fps
BN_800k                          244.897014   0.075914  13.172884
800k                             248.565521   0.077051  12.978469
faster_rcnn_R_50_FPN_3x          307.453097   0.095305  10.492657
BN_Mish_V3_80+_30k               311.967525   0.096704  10.340820
BN_Mish_V2_250+F_2_50k           314.006855   0.097336  10.273661
MTCNN                            329.380084   0.102102   9.794156
scratch_faster_rcnn_R_50_FPN_gn  332.273322   0.102999   9.708875

### COCO AP

Najlepszy wynik AP50 na FDDB osiagnął model 'faster_rcnn_R_50_FPN_3x' (91.43). AP dla wszystkich modeli za wyjątkiem '800k' mieściło się w zakresie od 37.6 do 39.45. 

'scratch_faster_rcnn_R_50_FPN_gn' trenowany od zera na zbiorze treningowy WIDER FACE uzyskał nejlepszy wynik na zbiorze validacyjnym datasetu. Model ma najwyższe AP50 = 91.43 przy AP = 39.45. Modele 'BN_Mish_V3_80+_30k' i 'BN_800k' mimo, że osiagneły wyższe wartości AP (kolejno 39.49 i 39.69) to ich AP50 nie przekroczył 90.5. Model osiagnał znacznie lepszy wynik 'faster_rcnn_R_50_FPN_3x' można wnioskować zatem o słuszności założeń prezentowanych w artykule "Rethinking ImageNet Pre-training" <a href="https://github.com/DarekGit/FACES_DNN/blob/master/notebooks/Bibliografia.ipynb">[15]</a>, które skutecznie udało się zastosować. 

Najlepsze wyniki na zbiorze FACES_DD uzyskał model 'BN_Mish_V2_250+F_2_50k'  	z AP = 62.25 i AP50 =	92.21. Pozostałe modele z wyjątkiem '800k' osiagneły porównywalne wyniki: AP w zakresie 58.02-60.86 i	AP50 w zakresie 88.98-90.89.

Otrzymane wyniki zostały zaprezentowane w poniższych tabelach.


Wynik dla FDDB


In [79]:
df_AP_FDDB

AP       AP50
faster_rcnn_R_50_FPN_3x          39.450135  91.435634
BN_Mish_V2_250+F_2_50k           37.604153  90.541335
BN_Mish_V3_80+_30k               39.490323  90.495200
BN_800k                          39.692626  89.397019
scratch_faster_rcnn_R_50_FPN_gn  37.811685  89.204056
800k                             19.019075  57.739784

Wynik dla WIDER FACE

In [81]:
df_AP_WIDERFACE

AP       AP50
scratch_faster_rcnn_R_50_FPN_gn  31.400140  57.145298
BN_Mish_V2_250+F_2_50k           30.202843  54.445577
BN_Mish_V3_80+_30k               29.157098  51.745406
faster_rcnn_R_50_FPN_3x          27.965053  50.158303
BN_800k                          27.766038  49.522419
800k                             23.100826  42.814369

Wynik dla FACES_DD

In [83]:
df_AP_FACES_DD

AP       AP50
BN_Mish_V2_250+F_2_50k           62.258940  92.213344
BN_Mish_V3_80+_30k               62.476118  91.280964
scratch_faster_rcnn_R_50_FPN_gn  58.026158  90.890998
BN_800k                          60.860485  89.135235
faster_rcnn_R_50_FPN_3x          59.622354  88.986689
800k                             47.816123  72.677722

### Ujednolicona miara oceny wyników detekcj mAP 

Na bazie FDDB najlepszy wyniki mAP osiągnał model 'BN_800k' uzyskujac mAP 40.33. Najwyższy mAP50 osiagnał model faster_rcnn_R_50_FPN_3x z wynikiem 92.22. Model '800k' ktory uzyskał najwyższy czas interferencji na zbiorze danych FDDB wypadł bardzo słabo (mAP = 19.07, mAP50 = 56.74). Moża zauważyć wpływ Batch normalization dzięki któremu model 'BN_800k' o porownywalnym czasem interferencji osiągnął znacznie lepszy wynik precyzji detekcji.

Na zbiorze WIDER FACE najlepszy wynik mAP uzyskał model 'scratch_faster_rcnn_R_50_FPN_gn' z mAP = 32.45 i mAP50 = 58.03. Fakt ten dowodzi skuteczności metody trenowania od zera na docelowym zbiorze treningowym z pomieciem procedury czasochłonnego pretrenowania na dużym zbiorze danych typu COCO.
'scratch_faster_rcnn_R_50_FPN_gn' osiagnał również lepszy wynik od 'faster_rcnn_R_50_FPN_3x' na zbiorze FACES_DD. Mimo porównywalnych wyników wypadł gorzej od modelu 'BN_Mish_V2_250+F_2_50k' z najwyższym mAP50 = 92.96 i mAP = 62.64.

Zauważyć można skuteczność ujednoliconej miara mAP. W wiekszości przypadków uzyskane wyniki były porównywalne z COCO AP. Zastosowanie tej miary pozwoliło również porównać wyniki detektora MTCNN, który na wszystich datasetach wypadł bardzo słabo uzyskując wyniki porównywalne z modelem '800k'

Otrzymane wyniki zostały zaprezentowane w poniższych tabelach.

Wynik dla FDDB

In [85]:
df_mAP_FDDB

mAP   All 0.50
faster_rcnn_R_50_FPN_3x          40.083175  92.229347
BN_Mish_V2_250+F_2_50k           38.287100  91.693780
BN_Mish_V3_80+_30k               40.209249  91.085791
scratch_faster_rcnn_R_50_FPN_gn  38.440094  90.080635
BN_800k                          40.337337  89.777419
MTCNN                            22.670603  76.760127
800k                             19.072028  56.749778

Wynik dla WIDER FACE

In [87]:
df_mAP_WIDERFACE

mAP   All 0.50
scratch_faster_rcnn_R_50_FPN_gn  32.450105  58.031785
BN_Mish_V2_250+F_2_50k           31.210473  55.337262
BN_Mish_V3_80+_30k               29.948815  52.611733
faster_rcnn_R_50_FPN_3x          28.727083  50.917348
BN_800k                          28.487464  50.312515
800k                             23.825164  43.351190
MTCNN                            13.332950  22.858543

Wynik dla FACES_DD

In [70]:
df_mAP_FACES_DD

mAP   All 0.50
BN_Mish_V2_250+F_2_50k           62.647906  92.960066
BN_Mish_V3_80+_30k               62.839073  91.776324
scratch_faster_rcnn_R_50_FPN_gn  58.432762  91.688343
BN_800k                          61.244431  89.821441
faster_rcnn_R_50_FPN_3x          60.223579  89.792963
800k                             48.101412  73.135868
MTCNN                            46.517802  62.864947

<br>

<br>

---

[Spis treści](https://github.com/DarekGit/FACES_DNN/blob/master/notebooks/Praca_Dyplomowa.ipynb) | [1. Wstęp](01_00_Wstep.ipynb) | [2. Metryki oceny detekcji](02_00_Miary.ipynb) | [3. Bazy danych](03_00_Datasety.ipynb) | [4. Przegląd metod detekcji](04_00_Modele.ipynb) | [5. Detekcja twarzy z wykorzystaniem wybranych architektur GSN](05_00_Modyfikacje.ipynb) | [6. Porównanie modeli](06_00_Porownanie.ipynb) | [7. Eksport modelu](07_00_Eksport_modelu.ipynb) | [8. Podsumowanie i wnioski](08_00_Podsumowanie.ipynb) | [Bibliografia](Bibliografia.ipynb)


---